### DISCUSSION TITLE
Define a Lisp implementation to run inside a (Guix) container

### DISCUSSION THREAD

**Discussion by Ambrevar** [joaotavora/sly#518] [Q&A] [answered]
I can start sbcl in a container with

```sh
guix environment --container --ad-hoc sbcl -- sbcl
```

But if I set the above as a `sly-lisp-implementations`, it tries connecting forever, to no avail.

What works however is to start the containerized Lisp as above in a separate shell, from there start a Slynk server with

```lisp
(slynk:create-server :port 4006)
```

Now from Emacs I can run `sly-connect` and it successfully connects to the containerized Lisp.

Any clue how to get `sly` to use a (Guix)-containerized Lisp directly?

**Comment by Ambrevar** [comment, marked as answer]
Answering my own question: the trick is to share the location of the SLY port file, which is where `(sly-slynk-port-file)` expands to, that is, `/tmp` on my machine.

This implementation works (bonus: extensions included):

```elisp
'("guix" "shell" "--container" "--network"
  "--expose=/gnu/store"
  "--share=/tmp"
  "sbcl" "sbcl-slynk"

  "emacs-sly"
  "emacs-sly-asdf"
  "emacs-sly-macrostep"
  "emacs-sly-named-readtables"
  "emacs-sly-stepper"
  "--" "sbcl")
```

@joaotavora Maybe document this somewhere?

**Comment by joaotavora** [comment]
I don't fully understand the question, but I am definitely happy you found the answer. 

> Maybe document this somewhere?

Maybe in your blog :-)  Else make a pull request to Sly's doc.  But then try to be as generic as possible, i.e. does this also work with Docker containers?

**Comment by Ambrevar** [comment]
I suppose it does.

What was blocking me was that I didn't know the existence of `sly-slynk-port-file` and that I had to share this file.

The solution is thus generic to any container system that supports sharing files... So all of them!
